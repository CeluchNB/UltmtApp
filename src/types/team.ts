export interface DisplayTeam {
    _id: string
    place: string
    name: string
    seasonStart: string
    seasonEnd: string
}

export interface CreateTeam {
    place: string
    name: string
    seasonStart: string
    seasonEnd: string
}
