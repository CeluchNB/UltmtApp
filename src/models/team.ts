import { DisplayUser } from '../types/user'
import { Realm } from '@realm/react'
import { LocalUser, Team } from '../types/team'

export const DisplayTeamSchema: Realm.ObjectSchema = {
    name: 'DisplayTeam',
    embedded: true,
    properties: {
        _id: 'string',
        place: 'string',
        name: 'string',
        teamname: 'string',
        seasonStart: 'string',
        seasonEnd: 'string',
    },
}

export const GuestTeamSchema: Realm.ObjectSchema = {
    name: 'GuestTeam',
    embedded: true,
    properties: {
        _id: 'string?',
        place: 'string?',
        name: 'string',
        teamname: 'string?',
        seasonStart: 'string?',
        seasonEnd: 'string?',
    },
}

export class TeamSchema {
    static schema: Realm.ObjectSchema = {
        name: 'Team',
        primaryKey: '_id',
        properties: {
            _id: 'string',
            place: 'string',
            name: 'string',
            teamname: 'string',
            managers: { type: 'list', objectType: 'DisplayUser' },
            players: { type: 'list', objectType: 'DisplayUser' },
            seasonStart: 'string',
            seasonEnd: 'string',
            seasonNumber: 'int',
            continuationId: 'string',
            rosterOpen: 'bool',
        },
    }

    _id: string
    place: string
    name: string
    teamname: string
    managers: DisplayUser[]
    players: LocalUser[]
    seasonStart: string
    seasonEnd: string
    seasonNumber: number
    continuationId: string
    rosterOpen: boolean

    constructor(team: Team) {
        this._id = team._id
        this.place = team.place
        this.name = team.name
        this.teamname = team.teamname
        this.managers = team.managers
        this.players = team.players
        this.seasonStart = team.seasonStart
        this.seasonEnd = team.seasonEnd
        this.seasonNumber = team.seasonNumber
        this.continuationId = team.continuationId
        this.rosterOpen = team.rosterOpen
    }
}
