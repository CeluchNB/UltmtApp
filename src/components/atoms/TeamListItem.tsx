import * as React from 'react'
import { DisplayTeam } from '../../types/team'
import { useColors } from '../../hooks'
import { StyleSheet, Text, View } from 'react-native'
import { size, weight } from '../../theme/fonts'

interface TeamListItemProps {
    team: DisplayTeam
}

const TeamListItem: React.FC<TeamListItemProps> = ({ team }) => {
    const { colors } = useColors()

    const styles = StyleSheet.create({
        teamName: {
            color: colors.gray,
            fontSize: size.fontMedium,
            fontWeight: weight.bold,
        },
        season: {
            color: colors.textPrimary,
            fontSize: size.fontSmall,
            fontWeight: weight.bold,
        },
    })

    return (
        <View>
            <Text style={styles.teamName}>{`${team.place} ${team.name}`}</Text>
            <Text style={styles.season}>{team.season}</Text>
        </View>
    )
}

export default TeamListItem
