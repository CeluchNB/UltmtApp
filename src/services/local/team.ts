import { Realm } from '@realm/react'
import { Team } from '../../types/team'
import { TeamSchema } from '../../models'
import { getRealm } from '../../models/realm'

export const saveTeams = async (teams: Team[]) => {
    const realm = await getRealm()
    realm.write(() => {
        for (const team of teams) {
            realm.create('Team', team, Realm.UpdateMode.Modified)
        }
    })
}

export const getManagingTeams = async (id: string): Promise<Team[]> => {
    const realm = await getRealm()
    const teams = await realm.objects<TeamSchema>('Team')
    return teams
        .filter(t => t.managers.map(m => m._id).includes(id))
        .map(t => parseTeam(t))
}

const parseTeam = (schema: TeamSchema): Team => {
    return JSON.parse(
        JSON.stringify({
            _id: schema._id,
            place: schema.place,
            name: schema.name,
            teamname: schema.teamname,
            managers: schema.managers,
            players: schema.players,
            seasonStart: schema.seasonStart,
            seasonEnd: schema.seasonEnd,
            seasonNumber: schema.seasonNumber,
            continuationId: schema.continuationId,
            rosterOpen: schema.rosterOpen,
            requests: [],
            games: [],
        }),
    )
}
