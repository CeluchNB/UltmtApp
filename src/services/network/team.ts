import { CreateTeam } from '../../types/team'
import { API_KEY, API_URL_V1 } from '@env'
import axios, { AxiosResponse } from 'axios'

export const createTeam = async (
    token: string,
    data: CreateTeam,
): Promise<AxiosResponse> => {
    return await axios.post(
        `${API_URL_V1}/team`,
        {
            team: { ...data },
        },
        {
            headers: {
                Authorization: `Bearer ${token}`,
                'X-API-Key': API_KEY,
            },
        },
    )
}

export const searchTeam = async (
    term: string,
    rosterOpen?: boolean,
): Promise<AxiosResponse> => {
    return await axios.get(
        `${API_URL_V1}/team/search?q=${term}&rosterOpen=${rosterOpen}`,
        {
            headers: { 'X-API-Key': API_KEY },
        },
    )
}

export const getManagedTeam = async (
    token: string,
    id: string,
): Promise<AxiosResponse> => {
    return await axios.get(`${API_URL_V1}/team/managing/${id}`, {
        headers: {
            Authorization: `Bearer ${token}`,
            'X-API-Key': API_KEY,
        },
    })
}

export const getTeam = async (id: string): Promise<AxiosResponse> => {
    return await axios.get(`${API_URL_V1}/team/${id}`, {
        headers: { 'X-API-Key': API_KEY },
    })
}

export const toggleRosterStatus = async (
    token: string,
    id: string,
    open: boolean,
): Promise<AxiosResponse> => {
    return await axios.put(
        `${API_URL_V1}/team/open/${id}?open=${open}`,
        {},
        {
            headers: {
                Authorization: `Bearer ${token}`,
                'X-API-Key': API_KEY,
            },
        },
    )
}

export const removePlayer = async (
    token: string,
    teamId: string,
    userId: string,
): Promise<AxiosResponse> => {
    return await axios.post(
        `${API_URL_V1}/team/remove/player/${teamId}?user=${userId}`,
        {},
        {
            headers: {
                Authorization: `Bearer ${token}`,
                'X-API-Key': API_KEY,
            },
        },
    )
}

export const rollover = async (
    token: string,
    teamId: string,
    copyPlayers: boolean,
    seasonStart: string,
    seasonEnd: string,
): Promise<AxiosResponse> => {
    return await axios.post(
        `${API_URL_V1}/team/rollover/${teamId}`,
        {
            copyPlayers,
            seasonStart,
            seasonEnd,
        },
        {
            headers: {
                Authorization: `Bearer ${token}`,
                'X-API-Key': API_KEY,
            },
        },
    )
}

export const addManager = async (
    token: string,
    teamId: string,
    managerId: string,
): Promise<AxiosResponse> => {
    return await axios.post(
        `${API_URL_V1}/team/${teamId}/addManager?manager=${managerId}`,
        {},
        {
            headers: {
                Authorization: `Bearer ${token}`,
                'X-API-Key': API_KEY,
            },
        },
    )
}

export const getArchivedTeam = async (
    teamId: string,
): Promise<AxiosResponse> => {
    return await axios.get(`${API_URL_V1}/archiveTeam/${teamId}`, {
        headers: {
            'X-API-Key': API_KEY,
        },
    })
}

export const createBulkJoinCode = async (
    token: string,
    teamId: string,
): Promise<AxiosResponse> => {
    return await axios.post(
        `${API_URL_V1}/team/getBulkCode?id=${teamId}`,
        {},
        {
            headers: {
                Authorization: `Bearer ${token}`,
                'X-API-Key': API_KEY,
            },
        },
    )
}
