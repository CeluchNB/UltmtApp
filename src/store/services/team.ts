import { ApiResponse } from '../../types/services'
import { CreateTeam } from '../../types/team'
import { unwrapResponse } from '../../utils/service-utils'

export const createTeam = async (
    token: string,
    data: CreateTeam,
): Promise<ApiResponse> => {
    const response = await fetch('https://ultmt-dev.herokuapp.com/team', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            Accept: '*/*',
            Authorization: `Bearer ${token}`,
        },
        body: JSON.stringify({
            team: {
                ...data,
            },
        }),
    })

    return unwrapResponse(response)
}

export const searchTeam = async (term: string): Promise<ApiResponse> => {
    if (term.length < 3) {
        return { error: 'Not enough characters to search' }
    }
    const response = await fetch(
        `https://ultmt-dev.herokuapp.com/team/search?q=${term}`,
    )
    return unwrapResponse(response)
}

export const getManagedTeam = async (
    token: string,
    id: string,
): Promise<ApiResponse> => {
    const response = await fetch(
        `https://ultmt-dev.herokuapp.com/team/managing/${id}`,
        {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
                Accept: '*/*',
                Authorization: `Bearer ${token}`,
            },
        },
    )

    return unwrapResponse(response)
}

export const toggleRosterStatus = async (
    token: string,
    id: string,
    open: boolean,
): Promise<ApiResponse> => {
    const response = await fetch(
        `https://ultmt-dev.herokuapp.com/team/open/${id}?open=${open}`,
        {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
                Accept: '*/*',
                Authorization: `Bearer ${token}`,
            },
        },
    )

    return await unwrapResponse(response)
}

export const removePlayer = async (
    token: string,
    teamId: string,
    userId: string,
): Promise<ApiResponse> => {
    const response = await fetch(
        `https://ultmt-dev.herokuapp.com/team/remove/player/${teamId}?user=${userId}`,
        {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                Accept: '*/*',
                Authorization: `Bearer ${token}`,
            },
        },
    )

    return await unwrapResponse(response)
}
