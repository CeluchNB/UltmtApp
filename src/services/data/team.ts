import * as Constants from '../../utils/constants'
import { ApiError } from '../../types/services'
import { throwApiError } from '../../utils/service-utils'
import { CreateTeam, Team } from '../../types/team'
import {
    createTeam as networkCreateTeam,
    getManagedTeam as networkGetManagedTeam,
    getTeam as networkGetTeam,
    removePlayer as networkRemovePlayer,
    rollover as networkRollover,
    searchTeam as networkSearchTeam,
    toggleRosterStatus as networkToggleRosterStatus,
} from '../network/team'

/**
 * Method to create team
 * @param token token of user creating team
 * @param data create team data
 * @returns created team
 * @throws error if team is not created successfully
 */
export const createTeam = async (
    token: string,
    data: CreateTeam,
): Promise<Team> => {
    try {
        const response = await networkCreateTeam(token, data)
        const { team } = response.data
        return team
    } catch (error) {
        return throwApiError(error, Constants.CREATE_TEAM_ERROR)
    }
}

/**
 * Method to search for teams by place, name, or username
 * @param term search term
 * @returns list of teams matching term
 * @throws error if backend returns an error
 */
export const searchTeam = async (term: string): Promise<Team[]> => {
    try {
        if (term.length < 3) {
            throw new ApiError('Not enough characters to search.')
        }
        const response = await networkSearchTeam(term)
        return response.data
    } catch (error) {
        return throwApiError(error, Constants.SEARCH_ERROR)
    }
}

/**
 * Method to get team that the current user manages
 * @param token jwt of current user
 * @param id team id
 * @returns team object
 * @throws error if backend returns error
 */
export const getManagedTeam = async (
    token: string,
    id: string,
): Promise<Team> => {
    try {
        const response = await networkGetManagedTeam(token, id)
        const { team } = response.data
        return team
    } catch (error) {
        return throwApiError(error, Constants.GET_TEAM_ERROR)
    }
}

/**
 * Method to get a public team
 * @param id id of team to get
 * @returns team object
 * @throws error if backend returns an error
 */
export const getTeam = async (id: string): Promise<Team> => {
    try {
        const response = await networkGetTeam(id)
        const { team } = response.data
        return team
    } catch (error) {
        return throwApiError(error, Constants.GET_TEAM_ERROR)
    }
}

/**
 * Method to change a team's roster status
 * @param token jwt of user changing team
 * @param id id of team to update
 * @param open boolean for team's roster status
 * @returns updated team object
 * @throws error if backend returns error
 */
export const toggleRosterStatus = async (
    token: string,
    id: string,
    open: boolean,
): Promise<Team> => {
    try {
        const response = await networkToggleRosterStatus(token, id, open)
        const { team } = response.data
        return team
    } catch (error) {
        return throwApiError(error, Constants.EDIT_TEAM_ERROR)
    }
}

/**
 * Method to remove a player from a team
 * @param token token of current user
 * @param teamId team player is on
 * @param userId id of player to remove
 * @returns updated team object
 * @throws error if backend returns an error
 */
export const removePlayer = async (
    token: string,
    teamId: string,
    userId: string,
): Promise<Team> => {
    try {
        const response = await networkRemovePlayer(token, teamId, userId)
        const { team } = response.data
        return team
    } catch (error) {
        return throwApiError(error, Constants.EDIT_TEAM_ERROR)
    }
}

/**
 * Method to start a new season for a team. After this method is called,
 * the old team is no longer editable.
 * @param token jwt of manager of team
 * @param teamId id of team to move over
 * @param copyPlayers boolean for deciding if current roster should be kept
 * @param seasonStart start year for season
 * @param seasonEnd start end for season
 * @returns newly created team object
 * @throws error if backend returns an error
 */
export const rollover = async (
    token: string,
    teamId: string,
    copyPlayers: boolean,
    seasonStart: string,
    seasonEnd: string,
): Promise<Team> => {
    try {
        const response = await networkRollover(
            token,
            teamId,
            copyPlayers,
            seasonStart,
            seasonEnd,
        )
        const { team } = response.data
        return team
    } catch (error) {
        return throwApiError(error, Constants.EDIT_TEAM_ERROR)
    }
}
