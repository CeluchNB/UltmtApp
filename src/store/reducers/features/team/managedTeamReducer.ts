import * as TeamData from '../../../../services/data/team'
import { RootState } from '../../../store'
import { Team } from '../../../../types/team'
import {
    RemovePlayerData,
    TeamRequestData,
    ToggleRosterStatusData,
} from '../../../../types/team'
import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'

export interface ManagedTeamSlice {
    team?: Team
    openLoading: boolean
    teamLoading: boolean
    error?: string
}

const initialState: ManagedTeamSlice = {
    team: undefined,
    openLoading: false,
    teamLoading: false,
    error: undefined,
}

const managedTeamSlice = createSlice({
    name: 'managedTeam',
    initialState,
    reducers: {
        setTeam(state, action) {
            state.team = action.payload
        },
        setTeamLoading(state, action) {
            state.teamLoading = action.payload
        },
    },
    extraReducers: builder => {
        builder
            .addCase(getManagedTeam.pending, state => {
                state.teamLoading = true
                state.error = undefined
            })
            .addCase(getManagedTeam.fulfilled, (state, action) => {
                state.teamLoading = false
                state.error = undefined
                state.team = action.payload
            })
            .addCase(getManagedTeam.rejected, (state, action) => {
                state.teamLoading = false
                state.error = action.error.message
            })
        builder
            .addCase(toggleRosterStatus.pending, state => {
                state.openLoading = true
                state.error = undefined
            })
            .addCase(toggleRosterStatus.fulfilled, (state, action) => {
                state.openLoading = false
                state.error = undefined
                state.team = action.payload
            })
            .addCase(toggleRosterStatus.rejected, (state, action) => {
                state.openLoading = false
                state.error = action.error.message
            })
        builder
            .addCase(removePlayer.pending, state => {
                state.teamLoading = true
                state.error = undefined
            })
            .addCase(removePlayer.fulfilled, (state, action) => {
                state.teamLoading = false
                state.error = undefined
                state.team = action.payload
            })
            .addCase(removePlayer.rejected, (state, action) => {
                state.teamLoading = false
                state.error = action.error.message
            })
    },
})

export const getManagedTeam = createAsyncThunk(
    'managedTeam/getTeam',
    async (data: TeamRequestData, _thunkAPI) => {
        const { token, id } = data
        return await TeamData.getManagedTeam(token, id)
    },
)

export const toggleRosterStatus = createAsyncThunk(
    'managedTeam/toggleRosterStatus',
    async (data: ToggleRosterStatusData, _thunkAPI) => {
        const { token, id, open } = data
        return await TeamData.toggleRosterStatus(token, id, open)
    },
)

export const removePlayer = createAsyncThunk(
    'managedTeam/removePlayer',
    async (data: RemovePlayerData, _thunkAPI) => {
        const { token, id, userId } = data
        return await TeamData.removePlayer(token, id, userId)
    },
)

export const selectTeam = (state: RootState) => state.managedTeam.team
export const selectTeamLoading = (state: RootState) =>
    state.managedTeam.teamLoading
export const selectOpenLoading = (state: RootState) =>
    state.managedTeam.openLoading
export const selectTeamError = (state: RootState) => state.managedTeam.error
export const { setTeam, setTeamLoading } = managedTeamSlice.actions
export default managedTeamSlice.reducer
