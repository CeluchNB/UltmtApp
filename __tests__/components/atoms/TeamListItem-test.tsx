import * as React from 'react'
import { DisplayTeam } from '../../../src/types/team'
import TeamListItem from '../../../src/components/atoms/TeamListItem'
import renderer from 'react-test-renderer'
import { fireEvent, render } from '@testing-library/react-native'

const team: DisplayTeam = {
    _id: 'id1',
    place: 'Place1',
    name: 'Name1',
    teamname: 'teamname1',
    seasonStart: '2019',
    seasonEnd: '2019',
}

it('matches snapshot', () => {
    const snapshot = renderer.create(<TeamListItem team={team} />)

    expect(snapshot).toMatchSnapshot()
})

it('displays team in same year', () => {
    const { getByText } = render(<TeamListItem team={team} />)

    expect(getByText(`${team.place} ${team.name}`)).toBeTruthy()
    expect(getByText(team.seasonStart)).toBeTruthy()
})

it('displays team in different year', () => {
    team.seasonEnd = '2020'
    const { getByText } = render(<TeamListItem team={team} />)

    expect(getByText(`${team.place} ${team.name}`)).toBeTruthy()
    expect(getByText(`${team.seasonStart} - ${team.seasonEnd}`)).toBeTruthy()
})

it('test on press', () => {
    const onPress = jest.fn()
    const { getByText, getByTestId } = render(
        <TeamListItem team={team} onPress={onPress} />,
    )
    const userDisplay = getByText('Place1 Name1')
    fireEvent.press(userDisplay)

    const goButton = getByTestId('go-button')
    fireEvent.press(goButton)
    expect(onPress).toHaveBeenCalledTimes(2)
})
