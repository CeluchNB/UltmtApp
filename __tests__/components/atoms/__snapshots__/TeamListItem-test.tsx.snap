// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<View
  style={
    Object {
      "display": "flex",
      "flexDirection": "row",
      "marginBottom": 5,
    }
  }
>
  <View
    style={
      Object {
        "alignSelf": "center",
        "flex": 1,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={false}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#707070",
            "fontSize": 41.0958904109589,
            "fontWeight": "bold",
          }
        }
      >
        Place1 Name1
      </Text>
      <Text
        style={
          Object {
            "color": "#3183ff",
            "fontSize": 41.0958904109589,
            "fontWeight": "bold",
          }
        }
      >
        @
        teamname1
      </Text>
      <Text
        style={
          Object {
            "color": "#3183ff",
            "fontSize": 20.54794520547945,
            "fontWeight": "bold",
          }
        }
      >
        2019
      </Text>
    </View>
  </View>
</View>
`;
